### ISSUE TITLE
Fine-tuning StaticEmbedding model with multilingual dataset is unexpectedly slow

### ISSUE THREAD

**Issue by NiklasUllmannDD** [UKPLab/sentence-transformers#3281]
I'm observing unexpected performance degradation when fine-tuning a StaticEmbedding model on a multilingual dataset compared to a single-language dataset. Training time increases dramatically beyond what would be expected from the dataset size increase alone.

## Details

- Single language: Training on only the German dataset takes ~0.5 hours for one epoch
- All languages: Training on all 7 languages takes ~24 hours for one epoch
- Expected behavior: Given the relative sizes of the datasets, I would expect a roughly linear increase in training time
Actual behavior: The training time for the multilingual dataset is disproportionately longer

## Environment

- StaticEmbedding model: `intfloat/multilingual-e5-large`
- Training infrastructure: Nvidia Tesla T4
- Training and code inspired by issue: #2982 

## Questions

Is this performance degradation expected when using multiple languages?
Are there any recommended optimizations for multilingual fine-tuning?
Could there be a specific bottleneck in the processing pipeline when handling multiple languages?

## Additional Information

Thats my dataset dict structure:
```
{'de': Dataset({
    features: ['anchor', 'positive'],
    num_rows: 5731565
}), 'pl': Dataset({
    features: ['anchor', 'positive'],
    num_rows: 545976
}), 'ch_fr': Dataset({
    features: ['anchor', 'positive'],
    num_rows: 236071
}), 'sk': Dataset({
    features: ['anchor', 'positive'],
    num_rows: 468963
}), 'hu': Dataset({
    features: ['anchor', 'positive'],
    num_rows: 463327
}), 'tr': Dataset({
    features: ['anchor', 'positive'],
    num_rows: 305002
}), 'cz': Dataset({
    features: ['anchor', 'positive'],
    num_rows: 705849
})}
```

Thats the code I train with:

```

    BATCH_SIZE =  2**13
    NUM_TRAIN_EPOCHS = 1
    batches_count = math.ceil(len(train_dataset) / BATCH_SIZE)
    eval_steps = int(batches_count * 0.2)
    warmup_steps = int(batches_count * NUM_TRAIN_EPOCHS * 0.1)


    args = SentenceTransformerTrainingArguments(
        output_dir="./tmp/model_train",
        num_train_epochs=NUM_TRAIN_EPOCHS,
        per_device_train_batch_size=BATCH_SIZE,
        per_device_eval_batch_size=BATCH_SIZE,
        learning_rate= best_parameters.get("learning_rate", 2e-1),
        batch_sampler=BatchSamplers.NO_DUPLICATES,
        eval_strategy="steps",
        eval_steps=eval_steps,
        warmup_steps=warmup_steps,
        bf16=True
    )
    static_embedding = StaticEmbedding.from_distillation(
        "intfloat/multilingual-e5-large",
        device=device,
        pca_dims=256,
        apply_zipf=True,
    )
    model = SentenceTransformer(modules=[static_embedding]).to(device)
    loss = MultipleNegativesRankingLoss(
        model,
        scale=50,
    )
    trainer = SentenceTransformerTrainer(
        model=model,
        args=args,
        loss=loss,
        train_dataset=train_dataset,
        eval_dataset=eval_dataset,
    )
    trainer.train()
```

Any insights on resolving this unexpected performance issue would be greatly appreciated. 
Great work! 🤗

**Comment by tomaarsen** [comment]
Hello!

Based on your dataset statistics, I indeed think the issue must be related to data preprocessing/loading before feeding to the model. This is not a terribly uncommon bottleneck for static models: the model part is extremely fast after all. I would do some simple experimentation with:
* [dataloader_num_workers](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.TrainingArguments.dataloader_num_workers) (e.g. 4)
* [dataloader_prefetch_factor](https://huggingface.co/docs/transformers/main/en/main_classes/trainer#transformers.TrainingArguments.dataloader_prefetch_factor)

If that still doesn't help, then I would also consider concatenating all input datasets, but you do get different performance: then you no longer only have in-batch negatives from the same language in `MultipleNegativesRankingLoss`.

Beyond that, I would keep an eye on your GPU usage over time. In my experience it generally starts a bit slow/low, but it should pick up to be high fairly consistently. If it's low, it's because the GPU isn't fed data fast enough, which is a pretty tricky issue to improve.

- Tom Aarsen

**Comment by NiklasUllmannDD** [comment]
Thanks @tomaarsen for the quick response!

I tested the approach with the dataloader settings and did not notice any big performance gains.
Next thing I will test is using one big dataset as you mentioned instead of 7 smaller ones.

Regarding the GPU usage: With my current config I see a big usage spike right at the beginning, probably, when I create the static embedding model. After that my GPU usage "idles" at around 4-5% during training with some spikes up to 15%. I am not observing any increasing usage, as you mentioned.

**Comment by aribornstein** [comment]
@tomaarsen I am also seeing this behavior on a very similar training script on Azure.

**Comment by tomaarsen** [comment]
I ran some experiments locally with arbitrary datasets:

<details><summary>My experimental training script</summary>

```python
import random
import logging
from datasets import concatenate_datasets, load_dataset, Dataset, DatasetDict
from torch import concatenate
from sentence_transformers import (
    SentenceTransformer,
    SentenceTransformerTrainer,
    SentenceTransformerTrainingArguments,
    SentenceTransformerModelCardData,
)
from sentence_transformers.losses import MatryoshkaLoss, MultipleNegativesRankingLoss
from sentence_transformers.training_args import BatchSamplers, MultiDatasetBatchSamplers
from sentence_transformers.models.StaticEmbedding import StaticEmbedding
from nltk.corpus import words

from transformers import AutoTokenizer

logging.basicConfig(
    format="%(asctime)s - %(message)s", datefmt="%Y-%m-%d %H:%M:%S", level=logging.INFO
)
random.seed(12)


def load_train_eval_datasets():
    """
    Either load the train and eval datasets from disk or load them from the datasets library & save them to disk.

    Upon saving to disk, we quit() to ensure that the datasets are not loaded into memory before training.
    """
    try:
        train_dataset = DatasetDict.load_from_disk("datasets/train_dataset")
        return train_dataset
    except FileNotFoundError:
        wordlist = words.words()
        
        dataset_size = 1_000_000
        dataset_names = ["dataset_A", "dataset_B", "dataset_C", "dataset_D", "dataset_E"]
        datasets = {}
        for dataset_name in dataset_names:
            # Create a dummy dataset with random sentences
            sentence_As = [" ".join(random.sample(wordlist, k=30)) for _ in range(dataset_size)]
            sentence_Bs = [" ".join(random.sample(wordlist, k=30)) for _ in range(dataset_size)]
            dataset = Dataset.from_dict({"sentence_A": sentence_As, "sentence_B": sentence_Bs})
            datasets[dataset_name] = dataset

        train_dataset = DatasetDict(datasets)
        train_dataset.save_to_disk("datasets/train_dataset")
        
        # The `train_test_split` calls have put a lot of the datasets in memory, while we want it to just be on disk
        quit()

def main():
    # 1. Load a model to finetune with 2. (Optional) model card data
    static_embedding = StaticEmbedding(AutoTokenizer.from_pretrained("google-bert/bert-base-multilingual-uncased"), embedding_dim=1024)
    model = SentenceTransformer(
        modules=[static_embedding],
        model_card_data=SentenceTransformerModelCardData(
            license="apache-2.0",
            model_name="Static Embeddings with BERT Multilingual uncased tokenizer finetuned on various datasets",
        ),
    )

    # 3. Set up training & evaluation datasets - each dataset is trained with MNRL (with MRL)
    train_dataset = load_train_eval_datasets()
    # 1 small dataset:
    # train_dataset = train_dataset["dataset_A"]
    # 1 big dataset:
    # train_dataset = concatenate_datasets(train_dataset.values())
    print(train_dataset)

    # 4. Define a loss function
    loss = MultipleNegativesRankingLoss(model)
    # loss = MatryoshkaLoss(model, loss, matryoshka_dims=[32, 64, 128, 256, 512, 1024])

    # 5. (Optional) Specify training arguments
    run_name = "static-similarity-mrl-multilingual-v1"
    args = SentenceTransformerTrainingArguments(
        # Required parameter:
        output_dir=f"models/{run_name}",
        # Optional training parameters:
        num_train_epochs=1,
        per_device_train_batch_size=2048,
        per_device_eval_batch_size=2048,
        learning_rate=2e-1,
        warmup_ratio=0.1,
        fp16=False,  # Set to False if you get an error that your GPU can't run on FP16
        bf16=True,  # Set to True if you have a GPU that supports BF16
        batch_sampler=BatchSamplers.NO_DUPLICATES,  # MultipleNegativesRankingLoss benefits from no duplicate samples in a batch
        multi_dataset_batch_sampler=MultiDatasetBatchSamplers.PROPORTIONAL,
        # Optional tracking/debugging parameters:
        eval_strategy="no",
        save_strategy="no",
        logging_steps=50,
        logging_first_step=True,
        run_name=run_name,  # Will be used in W&B if `wandb` is installed
    )

    # 6. Create a trainer & train
    trainer = SentenceTransformerTrainer(
        model=model,
        args=args,
        train_dataset=train_dataset,
        # eval_dataset=eval_dataset,
        loss=loss,
    )
    trainer.train()

    # 7. Save the trained model
    model.save_pretrained(f"models/{run_name}/final")

    # 8. (Optional) Push it to the Hugging Face Hub
    # model.push_to_hub(run_name, private=True)

"""
5 datasets:
00:50<00:00

{'train_runtime': 529.538, 'train_samples_per_second': 9442.193, 'train_steps_per_second': 4.617, 'train_loss': 7.631526069017032, 'epoch': 1.0}             
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2445/2445 [08:40<00:00,  4.70it/s]

1 big dataset:
00:43<00:00

{'train_runtime': 463.3478, 'train_samples_per_second': 10791.031, 'train_steps_per_second': 5.27, 'train_loss': 7.633710475752922, 'epoch': 1.0}            
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2442/2442 [07:36<00:00,  5.35it/s]

1 small dataset:
{'train_runtime': 89.423, 'train_samples_per_second': 11182.802, 'train_steps_per_second': 5.468, 'train_loss': 7.643406126884352, 'epoch': 1.0}             
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 489/489 [01:22<00:00,  5.94it/s]
"""

if __name__ == "__main__":
    main()
```

</details>

Here, I created 5 separate datasets of 1 million text pairs, and I saved them on disk at one big DatasetDict. Here are my results:
* 5 datasets, 1 million text pairs each: 08:40 training time (529.538 seconds)
* 1 dataset, 5 million text pairs via concatenation: 07:36 training time (463.3478 seconds)
* 1 dataset, 1 million text pairs via only indexing (only selecting the first dataset): 01:22 training time (89.423 seconds)

So: 
1. There is definitely a bit of an overhead for using 5 datasets compared to 1 big dataset, about 14%, but it's not outrageous.
2. Using just 1 dataset (e.g. selecting the first dataset) compared to the biggest concatenated dataset takes about 5.18x less time than the 1 dataset with 5 million text pairs. This isn't too weird either.

And crucially: the ratio of 1 dataset with 1 million vs 5 datasets with 1 million each is 1.18x if we ignore the difference in total text pairs. This is roughly linear, which clashes with your findings which did not seem linear at all. I also used `BatchSamplers.NO_DUPLICATES` as that could have been a reason.

In short, I'm struggling to properly reproduce this with arbitrary/fake datasets, sadly.

- Tom Aarsen

**Comment by talbaumel** [comment]
I had an issue training static embeddings, due to the very large batch size and no duplications sampling on some of my datasets it was very hard to find batches.

**Comment by tomaarsen** [comment]
It's certainly possible that the `BatchSamplers.NO_DUPLICATES` is partially responsible, perhaps depending on your CPU.

- Tom Aarsen

**Comment by NiklasUllmannDD** [comment]
Thank you @tomaarsen, @talbaumel and @aribornstein  for your patience as we thoroughly investigated the training performance issues. After extensive testing with various parameter combinations and implementing detailed monitoring, we've identified the root cause: 

**it's not slow training, but extremely slow evaluation**

# Monitoring Setup

Implemented TensorBoard callback to track and compare training metrics across runs
Added hardware monitoring via Vertex AI pipeline components
Current configuration:

1 epoch training with logging every 5% (20 steps total)
Evaluation runs every 20% (5 evaluations per epoch)
70/30 train/evaluation split



## What We've Tested So Far
1. Batch Size Adjustments

- Systematically reduced batch size from 2^12 down to 2^4
- Hypothesis: Large batches with NO_DUPLICATES setting might cause inefficiencies
- Result: No significant performance improvement

2. Mixed Precision (FP16)

- Implemented FP16 training on our Tesla T4 GPU
- Result: Noticeable performance improvement for training but not huge gains

3. CPU & DataLoader Optimizations

- Initially: No performance gains with suggested DataLoader settings
- After increasing CPU allocation: Visible training performance improvement

## Current Hardware Utilization

GPU: Tesla T4 (16GB VRAM)
- Computation: Only 3% utilization during training, spikes to 12% during evaluation
- Memory: Consistently at 80% VRAM utilization


System: 32GB RAM, multi-core CPU
- RAM: ~50% utilization during both training and evaluation
- CPU: 10-30% utilization



## The Bottleneck Identified

- Training segments process quickly (~6 minutes per 5% of an epoch)
- Evaluation segments are extremely slow (~3 hours each)


## Follow up questions:
Do you have any guesses or suggestions why  the evaluation is taking so much longer than training and how can this time be reduced at all?

- Reducing the amount of times we evaluate during an epoch is the first option, but not likely the best, since we need those information.
- As a long term goal we would like to use a custom evaluator, maybe that will help with our evaluation speeds.

**Comment by tomaarsen** [comment]
Thank you for sharing! In Sentence Transformers you can perform evaluation in 2 ways:
* `eval_dataset`: An evaluation dataset in the same format as your training dataset. This will be used with your loss functions to get an evaluation loss. Useful to avoid overfitting, but not much else.
* `evaluator`: A `SentenceEvaluator` instance whose `__call__` methods gets called on every evaluation, at which point the evaluator can do whatever it wants, as long as it returns some metrics dictionary.
You can use any combination of these 2.

As far as I can tell, based on your script above, you're only using the `eval_dataset`. Normally, one step with evaluation is *faster* than one step with training, as it's nearly identical minus that we don't need to 1) backward or 2) track gradients. So, if your evaluation is very very slow, then my expectation is that your `eval_dataset` is extremely large.

Because the `eval_dataset` primarily exists to ensure that you're not overfitting, you can very reasonably cut it way down. For example, you can use e.g. a `DatasetDict` or `dict` that maps dataset names to a `Dataset` with e.g. 100 to 1000 samples. Out of curiosity, how big is your `eval_dataset` currently?

- Tom Aarsen

**Comment by NiklasUllmannDD** [comment]
Hey @tomaarsen ,

Yes, our evaluation dataset is about 30% of our total dataset, as we are used to from other machine learning. Here are the specific values:

- Train Dataset:
`{'de': 5734037, 'pl': 544802, 'ch_fr': 236219, 'sk': 469866, 'hu': 463398, 'tr': 305411, 'cz': 705734}`

- Eval Dataset:
`{'de': 2464084, 'pl': 235576, 'ch_fr': 102058, 'sk': 201236, 'hu': 199683, 'tr': 130887, 'cz': 302078}`



Yes, thats way beyond the 1000 examples you proposed. I can follow your arguments why evaluation should be faster than training, that also reflects my understanding of it. But I guess that in the evaluation step there might be a bottleneck somewhere, maybe between moving data around or elsewhere. 

Next steps for our testing is to reduce the eval_dataset size and also implement an evaluator with the full eval_dataset as a test. 

Thank you very much, I will get back to you with updates!

**Comment by tomaarsen** [comment]
30% is indeed super reasonable in most ML tasks, but we indeed have to be mindful of what this means for our total training time. If we take a simple example: 1 dataset with 1 million samples, and we take a 70-30% train-eval split. We evaluate 20 times in total, i.e. every 5% of training, then:
1. We train for the first 5%, i.e. 35.000 samples.
2. Evaluate on the evaluation set, i.e. 300.000 samples.
3. Repeat 20x

In the end, we've used 6.700.000 samples, of which only 700.000 a.k.a. ~10.4% was used for training.

If you'd like to keep using 30% for evaluation, then I would recommend instead a 69-1-30 train-eval-test split or something like that. Then you use much less samples every 5% of training, you can still see if there's overfitting (if the eval loss is falling alongside the train loss) and then you can use the 30% test set after training.

But personally I would use a 90-1-9 train-eval-test split.

- Tom Aarsen

**Comment by NiklasUllmannDD** [comment]
Thank you @tomaarsen for your guidance throughout this investigation!

I'm closing this issue with two important discoveries that could benefit others in the community:

## Key Learnings:
### Evaluation Dataset Proportions Matter Significantly
- The 90-1-9 train-eval-test split you recommended is vastly more efficient than our initial 70-30 approach
- When evaluating frequently during training (every 5% in our case), a large evaluation dataset creates a severe performance bottleneck
- This insight should definitely be added to the documentation as it's not immediately obvious, especially to those coming from other ML frameworks
### Hardware Utilization 
- Despite having a Tesla T4 GPU available, the GPU was barely utilized at training and evaluation.
- Training on CPUs proved significantly faster in our specific use case.
